dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,205 @@
-type DevicePoolSpec struct {
-	// NodeName is the name of the node containing the devices in the pool.
-	// For network attached devices, this may be empty.
+type ResourcePool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
 	// +optional
-	NodeName *string `json:"nodeName,omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	// Driver is the name of the DeviceDriver that created this object and
-	// owns the data in it.
-	// +required
-	Driver string `json:"driver,omitempty"`
-
-	// Attributes contains device attributes that are common to all devices
-	// in the pool.
-	// +optional
-	// +listType=atomic
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResources are pooled resources that are shared by all devices
-	// in the pool. This is typically used when representing a
-	// partitionable device, and need not be populated otherwise.
+	// NodeName identifies the node which provides the devices
+	// if (and only if) they are local to a node. If this is empty,
+	// the devices exist outside of any specific node.
+	//
+	// NodeName and NodeSelector are mutually exclusive. If none is
+	// set, the devices are available in the entire cluster.
+	//
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain node name.
 	//
 	// +optional
-	// +listType=atomic
-	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,2,opt,name=nodeName"`
 
-	// Devices contains the individual devices in the pool.
+	// NodeSelector identifies all nodes that devices which are shared
+	// between nodes could be accessed from.
 	//
-	// +required
-	// +listType=atomic
-	Devices []Device `json:"devices,omitempty"`
-}
+	// NodeName and NodeSelector are mutually exclusive. If none is
+	// set, the devices are available in the entire cluster.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty"`
 
-// DevicePoolStatus contains the state of the pool as last reported by the
-// driver. Note that this will not include the allocations that have been made
-// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
-// to make future scheduling decisions.
-type DevicePoolStatus struct {
-	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
-}
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
 
-// AllocatedDevice represents a device that has been allocated from the pool.
-type AllocatedDevice struct {
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	ResourceModel `json:",inline" protobuf:"bytes,4,name=resourceModel"`
```

Comment:
I'm fine with renaming. It's not really part of the user-facing API either.

"PoolResources" sounds good to me.

Reply:
"Simple" just doesn't sound right, no matter where we put it.

I'm okay with renaming "NamedDevices" to just "Devices". That way the API looks "simple" (sic!) but we don't loose the ability to extend.

Side note:
```
ResourcePool {
    Simple: [ // this is a one-of, could be "simple", "partitioned", "compound", "pooled", ...
        { Name, Attributes }...
    ]
}
```

This doesn't work, entries in a one-of cannot be slices.
```
type OneOf struct {
   IntSlice []int
   StringSlice []string
}
```

Here we cannot tell which slice is set. JSON and protobuf do not distinguish between empty and null reliably.

This would be better, but go-protobuf generates broken code (I tried!):
```
type OneOf struct {
   IntSlice *[]int
   StringSlice *[]string
}
```

Reply:
What about using the actual name `OneOf` at the top level as a replacement for `ResourceModel`, i.e.:
```
ResourcePool {
    OneOf {
        <NEW NAME FOR THIS SINCE WE PLAN TO HAVE SHARED RESOURCES NEXT TO DEVICES UNDER HERE> {
            Devices: [
                { Name, Attributes }...
            ]
        }
    }
}
```

Reply:
That works technically. I'm not sure whether it is better, and we'd still need to come up with a `<NEW NAME FOR THIS ...` for the current model.

Reply:
> entries in a one-of cannot be slices.

Who says?  Unless a specified-but-empty list needs to be distinguished from unspecified?  Either way, I think this is probably not he best structure.

> I'm okay with renaming "NamedDevices" to just "Devices"

This makes the "simple" approach be `pool.devices.devices` and any future extensions become `pool.compound.devices` or `pool.partitioned.devices` which isn't SO bad.

Do we really need this one of?  Are we fairly sure that we can't model the other kinds of devices in the same struct?

Reply:
> Who says? Unless a specified-but-empty list needs to be distinguished from unspecified?

In a one-of-many, we must use `omitempty`. But JSON then omits both a zero-length slice and a null slice, so it becomes impossible to distinguish between "empty" and "unset". A pointer to a slice doesn't work (go-protobuf code generator issue).

> Do we really need this one of? Are we fairly sure that we can't model the other kinds of devices in the same struct?

I'd like to turn the argument around: are we sure that we *can* model the other kinds of devices in the same struct?

I'm not sure, therefore I'd like to have this extension point.

Reply:
Would it make sense to pull in all those different proposals and put them into the PR for another round of reviews?

Reply:
Actually, I do need to update capacity_types.go. I thought I had removed the more complex model for MIG, but it's still there.

Reply:
Spelling out the actual proposal helped a bit. Perhaps it has been mentioned before, but we don't really need to limit a ResourcePool to just one kind. There's nothing wrong with mixing "Devices" and future "Gizmos" in the same pool if they share the other fields (driver name, node name).

Dropping the one-of works in this case because a pool with no known field populated is simply an empty pool for old clients. That works in this case because it is safe for them to ignore it.

Reply:
I think it benefits the discussion to share that update. See https://github.com/kubernetes-sigs/wg-device-management/pull/14/commits/7aeb3c49d140e9562fe993d0c1177ba8bccce2a7

I converted one pool example and it has exactly the same structure as in k8srm-prototype (see dra-evolution/testdata/pools-one-node-sriov-0.yaml).

Reply:
I do think that is better. But I am unclear on why we need to have a generic "ResourcePool" rather than several different API kinds in the same API group, representing the different types of things that can be managed? I know that additional top-level types are discouraged, but it seems when we are modeling very different things, we should use different types. What do we *gain* by combining the type? Compared to something that eventually looks like:

- `NodeDevices` or `NodeDevicePool` - one entry per driver, per node, looks like our simple device model. Maybe this can also cover partitionable devices, may not.
- `NetworkDevices` or `NetworkDevicePool` - one entry per driver per set of managed network-attachable devices. Maybe this can also cover disaggregated devices since they have the same reachability and lifecycle requirements.
- `FutureCrazyDevicePool` - we have no conception of this, but we don't have to...it can have totally different semantics

They might share some common types, like `DeviceAttributes`, so that the claim model can rely on those being available.

Reply:
> What do we gain by combining the type?

All code that doesn't care about what's inside the pool, like for example permission checks and the entire boilerplate code for managing the type, has to be replicated. If a user wants to know about DRA resources, they have to remember which types are available in which Kubernetes version and check all of them. Admins have to remember to adapt RBAC rules for their users when we add a new type.

The difference between network-attached and local devices is smaller than you seem to think. That doesn't justify an entirely different type that just differs by one field.

Reply:
Defining a new type should be the last resort (it's still an option!), not the go-to solution.

Reply:
Another argument could become that "devices" and "gizmos" might need to share the same consumable resources in a pool. This doesn't work if they have to be in different types, at least not with the "consumable resources" definition that @klueska shared where those resources are only available to devices in the same pool.

This didn't work with the one-of that I had before, but now it would be possible to list "devices" and "gizmos" together in that same pool.

Reply:
> The difference between network-attached and local devices is smaller than you seem to think. That doesn't justify an entirely different type that just differs by one field.

Could be. The semantics and lifecycle are really, really different though, even if the fields aren't. The reachability is different, the cardinality is different, and their is a provisioning lifecycle that is different.

Reply:
>> What do we gain by combining the type?

> All code that doesn't care about what's inside the pool, like for example permission checks and the entire boilerplate code for managing the type, has to be replicated. If a user wants to know about DRA resources, they have to remember which types are available in which Kubernetes version and check all of them. Admins have to remember to adapt RBAC rules for their users when we add a new type.

Replicating boilerplate code for managing the types is not that big of a deal to me. However, the second (and third) part of Patrick's argument is where he won me over last time we discussed this.

Reply:
None of those arguments convince me. Boilerplate is what it is, that's just an inconvenience. Kubernetes could have created a generic "Workload" kind and had StatefulSet, Deployment, and DaemonSet be one-ofs off of that. It would have been terrible. I see this as pretty similar.

The second and third arguments don't really make sense to me. I don't think user's will think in terms of "DRA resources". They will think of "resources" or maybe "devices". If a user wants know about a networked device, rather than an on-node device, I don't think it will be a problem for them to look at a different kind of pool. In fact they may well expect to; not being associated with a particular node is a really different thing that devices that are installed directly in nodes.

Default RBAC rules will be installed, like other kinds. The admin having a little *more* control and being able to differentiate between on-node and networked devices does not seem like a downside. Again, the lifecycle and management of networked pools vs on-device pools are different, so their access control needs  may be different.

Reply:
Given that this thread is attached to a fundamentally changed line of code, it's hard to reply.  It's hard to even know what we are discussing, really.

Here's my understanding of the question at hand:  Should a `<Something>Pool` contain different kinds of `Somethings`?

The analogy with workloads doesn't feel quite right to me.  Deployment and StatefulSet are fundamentally unalike wrt API, semantics, and use-cases.  Deployment and DaemonSet *could* have been the same API object.  They started very different, but ultimately got closer to each other.  I don't know if it would have made a better API.  But that's moot because there's no place I know of where we have to process "all deployments and all daemonsets" at the same time.  

I am deeply skeptical that we will ever need a `Gizmo` that parallels `Device` but couldn't be referred to as a "Device", and so I think it reduces to:

```
ResourcePool {
  Name
  Spec {
    // Each device structure is internally capturing the different archetypes of device
    Devices: [ { complexity: Simple | Partitioned | CompoundDevice | ... }, ... ]
  }
}
```

vs. 

```
ResourcePool {
  Name
  Spec {
    // Each list carries a similar but not identical structure
    Devices[ { simple device }, ...]
    PartitionedDevices[ { partitioned device }, ...]
    CompoundDevices[ { compound device }, ...]
    ...
  }
}
```

vs.

```
// Each type carries a similar but not identical structure

DevicePool {
  Name
  Spec {
    Devices[ { simple device }, ...]
    ...
  }
}

PartitionedDevicePool {
  Name
  Spec {
    PartitionedDevices[ { partitioned device }, ...]
    ...
  }
}

CompoundDevicePool {
  Name
  Spec {
    CompoundDevices[ { compound device }, ...]
    ...
  }
}
```

This last option seems like overkill for Devices, IMO.  Of the previous 2, I can go either way.

Renaming ResourcePool to DevicePool seems, to me, logical and correct but it IS more work (we'd probably want to rename the apigroup?).

Reply:
The last three actually would all likely still be one type, in my thinking. The new type would come into play with networked devices. The cardinality with node is different. The lifecycle is different. The allocation logic is different.

Reply:
> The new type would come into play with networked devices. The cardinality with node is different. The lifecycle is different. The allocation logic is different.

None of those require a new type.

- The cardinality with node is different: Each object has one owner. For node-local, it's the node and the driver on that node. For network-attached, it is the driver controller.
- The lifecycle is different: Garbage-collection of objects is done through OwnerRef in both cases. Which devices are listed can change in both cases (hot-plugging, node reboots) and is determined by the owner.
- The allocation logic is different: In both cases the flow is: when filtering nodes, ignore pools which are not accessible from the node under evaluation (node name comparison resp. node selector evaluation), then check devices from pools that are accessible. The protocol for reserving devices when we allow competition is the same.

If it makes you feel better, then we can rename ResourcePool -> DevicePool "because it contains only devices". But I don't see the justification for renaming the resource.k8s.io API group and the pod spec fields.

Reply:
> None of those require a new type.
> 

It's not so much that I think they *require* it, it's that I think it would simplify the API and the things we have to take into consideration for that type.

> * The cardinality with node is different: Each object has one owner. For node-local, it's the node and the driver on that node. For network-attached, it is the driver controller.

That's not what I mean. What I mean is that node-local devices are only ever used by one node. So, we don't have to think about that other nodes might be consuming them. Networked devices are consumed from multiple nodes, and so there are different considerations. For example, we need the label selector to show us which nodes are usable with the device. I think the way we track allocations may also need to be different. We may need different quota and access control mechanisms. I just think we're adding a lot of additional design burden for something we don't need right away.

> * The lifecycle is different: Garbage-collection of objects is done through OwnerRef in both cases. Which devices are listed can change in both cases (hot-plugging, node reboots) and is determined by the owner.

That's not what I mean by lifecycle. I am talking about the fact that the device needs to be created and/or attached to the node or pod differently than a node-local device. There is a distributed handshake that has to happen between the node and the network service managing the networked device. This introduced a setup protocol that needs to be managed and different types of failures need to be accounted for, in a way we do not have with node-local devices. You also likely have a higher failure rate in these devices and so we may want to manage failures differently. All of that may require additional logic and maybe additional fields (in the spec or status).

> * The allocation logic is different: In both cases the flow is: when filtering nodes, ignore pools which are not accessible from the node under evaluation (node name comparison resp. node selector evaluation), then check devices from pools that are accessible. The protocol for reserving devices when we allow competition is the same.

I think we will find some more differences, for example reference counting access to network devices, etc. I think there will be different constraints and use cases.

I think it's overly optimistic to think these things are so similar. All that said, I am open to being wrong :)

> 
> If it makes you feel better, then we can rename ResourcePool -> DevicePool "because it contains only devices".

I think DevicePool is better but this is less critical than the decision discussed above (e.g., I would be OK with NodeResourcePool if we decided node-local and networked are different enough).

> But I don't see the justification for renaming the resource.k8s.io API group and the pod spec fields.

I agree, I think it makes sense to leave the group name the same, even if we change the kind names. Of course, I think you raised it before Patrick, in that case should it just be Pool? I think probably not, but you know, we don't have enough options yet so I figured I'd raise another!

Reply:
> I think it's overly optimistic to think these things are so similar. All that said, I am open to being wrong :)

Some discussions off the top of my head where the networked device use case has been raised already:
- Node selector discussion on pools
- Debate over uniqueness of device names
- Discussion on status and on where to store allocations

I think there were a few more. My point is that we are spending time on this now, and we haven't even gotten to the implementation, how to handle the provisioning lifecycle and coordination protocol for these, etc. I just think we're setting ourselves up for a lot of work that we could defer until later.